src/librustc/ty/mod.rs

```diff
@@ -2875,8 +2875,8 @@ impl<'tcx> TyCtxt<'tcx> {
                 _ => false,
             }
         } else {
-            match self.def_kind(def_id).expect("no def for `DefId`") {
-                DefKind::AssocConst | DefKind::AssocFn | DefKind::AssocTy => true,
+            match self.def_kind(def_id) {
+                Some(DefKind::AssocConst | DefKind::AssocFn | DefKind::AssocTy) => true,
```

Comment:
Now that this is fixed, maybe this special case can be removed: https://github.com/rust-lang/rust/blob/d1e81ef234ff5c2e0e3a69cb4e8e5f5b0fe1fd83/src/librustc/ty/mod.rs#L3057-L3063